### PR TITLE
Checking for length of categories before doing string conversion. fix…

### LINKED ISSUE #11305: PERF: rendering of large number of categories
xref #11304 

```
items = [str(i) for i in range(40000)]
s = pd.Series(items, dtype='category')
df = pd.DataFrame({'C':s, 'data':np.random.randn(40000)})
data = df[df.C=='20']
str(data.C)
```

This renders the string of all of the categories before chopping them, see [here](https://github.com/pydata/pandas/blob/master/pandas/core/categorical.py#L1392)

This should check for a large number of categories first, then render the chopped portions.

### PR DESCRIPTION
**Pull request by michaelaye**
closes #11305

### CODE REVIEW

**Comment by jreback** [comment]
- can you add a whats new note.
- pls add an asv benchmark (like your sample but only use 1000 categories)

**Comment by jreback** [comment]
can you update

**Comment by michaelaye** [comment]
Coming up. Had to read up where/how to do that.

**Comment by michaelaye** [comment]
the categorical asv benchmark file breaks PEP8 302: 2 lines between functions. Should I correct? I searched but did not find a least of PEP8 errors the pandas decided to ignore.

File: asv_bench/benchmarks/categoricals.py [hunk 1]
```diff
@@ -43,3 +45,16 @@ def time_regular_constructor(self):
...
+
+    def setup(self):
+        n = 1000
+        items = [str(i) for i in range(n)]
+        s = pd.Series(items, dtype='category')
+        df = pd.DataFrame({'C': s, 'data': np.random.randn(n)})
+        self.data = df[df.C == '20']
+
+    def time_rendering(self):
+        str(data.C)
```
Comment (jorisvandenbossche) [inline asv_bench/benchmarks/categoricals.py]: data -> self.data ?

**Comment by jorisvandenbossche** [comment]
@michaelaye that's ok for the PEP8 changes

File: doc/source/whatsnew/v0.17.1.txt [hunk 2]
```diff
@@ -55,7 +55,7 @@ Bug Fixes
 
 - Bug in ``.to_latex()`` output broken when the index has a name (:issue: `10660`)
 - Bug in ``HDFStore.append`` with strings whose encoded length exceded the max unencoded length (:issue:`11234`)
-
+- Performance bug in ``Categorical._repr_categories`` was rendering string before chopping them for display (:issue: `11305`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.17.1.txt]: move to Performance section

File: asv_bench/benchmarks/categoricals.py (hunk 1 above)
Comment (michaelaye) [inline asv_bench/benchmarks/categoricals.py]: Yeah, sorry. I can't run asv on my machine. first it complained about a missing config file, when i tried to run only the categorical bench test. apparently, when i'm not restricting it with a `-b` flag, it creates a new environment, but it seems to be incompatible with conda, as it creates a py2.7 environment now, while being run in a conda python3.4 environment.

Comment (michaelaye) [inline asv_bench/benchmarks/categoricals.py]: or maybe, that's meant to happen, as it tests for all to be tested environment? I don't know, have never used asv before.

**Comment by michaelaye** [comment]
I was able to start a run,  but for some reason, it only runs a python2.7 env, don't know why:

```bash
$ asv continuous master HEAD -b categorical
· Creating environments
· Discovering benchmarks
·· Uninstalling from py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-scipy-sqlalchemy-xlrd-xlwt
·· Building for py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-scipy-sqlalchemy-xlrd-xlwt...................................
·· Installing into py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-scipy-sqlalchemy-xlrd-xlwt..
· Running 12 total benchmarks (2 commits * 1 environments * 6 benchmarks)
[  0.00%] · For pandas commit hash cdff5bce:
[  0.00%] ·· Building for py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-scipy-sqlalchemy-xlrd-xlwt.........................................
[  0.00%] ·· Benchmarking py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-scipy-sqlalchemy-xlrd-xlwt
[  8.33%] ··· Running categoricals.categorical_constructor.time_fastpath                                4.85ms
[ 16.67%] ··· Running categoricals.categorical_constructor.time_regular_constructor                   226.53ms
[ 25.00%] ··· Running categoricals.categorical_rendering.time_rendering                                 2.31ms
[ 33.33%] ··· Running categoricals.categorical_value_counts.time_value_counts                          18.63ms
[ 41.67%] ··· Running categoricals.categorical_value_counts.time_value_counts_dropna                   24.74ms
[ 50.00%] ··· Running categoricals.concat_categorical.time_concat_categorical                          46.91ms
[ 50.00%] · For pandas commit hash c2aa6a23:
[ 50.00%] ·· Building for py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-scipy-sqlalchemy-xlrd-xlwt......................................
[ 50.00%] ·· Benchmarking py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-scipy-sqlalchemy-xlrd-xlwt
[ 58.33%] ··· Running categoricals.categorical_constructor.time_fastpath                                4.93ms
[ 66.67%] ··· Running categoricals.categorical_constructor.time_regular_constructor                   210.66ms
[ 75.00%] ··· Running categoricals.categorical_rendering.time_rendering                                17.63ms
[ 83.33%] ··· Running categoricals.categorical_value_counts.time_value_counts                          13.60ms
[ 91.67%] ··· Running categoricals.categorical_value_counts.time_value_counts_dropna                   17.59ms
[100.00%] ··· Running categoricals.concat_categorical.time_concat_categorical                          50.98ms    before     after       ratio
  [c2aa6a23] [cdff5bce]
-   17.63ms     2.31ms      0.13  categoricals.categorical_rendering.time_rendering
SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
```

File: doc/source/whatsnew/v0.17.1.txt [hunk 3]
```diff
@@ -50,6 +50,8 @@ Performance Improvements
 
 .. _whatsnew_0171.bug_fixes:
 
+- Performance bug in ``Categorical._repr_categories`` was rendering string before chopping them for display (:issue: `11305`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.17.1.txt]: move to Performance section

say ``Performance issue in rendering a large number of categories when printing a Categorical or Series of category dtype``

**Comment by jreback** [comment]
comments, then pls squash

**Comment by jreback** [comment]
merged via https://github.com/pydata/pandas/commit/47778004fbe2594d0f877fbf49b685ad55469557

thanks!

**Pull request closed without merging**